src/passes/function_passes/UnoptimizedCopyPass.cpp

```diff
@@ -64,20 +63,22 @@ struct UnoptimizedCopyPass: public FunctionPass {
             Function *calledFun = call.getCalledFunction();
             if (getPatchpointType(calledFun).isPatchpoint()) {
               Value *callback = call.getArgOperand(2)->stripPointerCasts();
-              Function *callbackFunction = cast<Function>(callback);
-              StringRef calledFunName = callbackFunction->getName();
-              if (calledFunName != GUARD_FUN &&
-                  !calledFunName.startswith(UNOPT_PREFIX)) {
-                // This is not an unoptimized function -> call the unoptimized
-                // version of this function instead.
-                Function *newFun =
-                  mod->getFunction(UNOPT_PREFIX + calledFunName.str());
-                if (newFun) {
-                  Type *i8ptr_t = PointerType::getUnqual(
-                      IntegerType::getInt8Ty(mod->getContext()));
-                  Constant* newCallback =
-                    ConstantExpr::getBitCast(newFun, i8ptr_t);
-                  call.setArgOperand(2, newCallback);
+              if (callback && isa<Function>(callback)) {
```

Comment:
So this look like the same code inside an extra condition. Can you explain the new condition?

Reply:
That is right. 

`calledFun` is an `llvm.experimental.patchpoint` call, and `callback` is its callback. `llvm.experimental.patchpoints` can take null pointers as callbacks. `CheckPointPass` inserts a `patchpoint` with a null callback in one of the `__unopt_` functions. This particular patch point corresponds to the patch point that called `__guard_failure` in the optimized version of the function. The purpose of the new condition is to avoid crashing when that call is processed (since you can't cast `nullptr` to `Function`).

Reply:
OK